xarray_jax/__init__.py

```diff
@@ -402,104 +354,78 @@ def jax_vars(
   """Like unwrap_vars, but will complain if vars are not all jax arrays."""
   return cast(Mapping[str, jax.Array], unwrap_vars(dataset, require_jax=True))
 
+class ScalarLeafWrapper:
+  """Wraps non-array data into a duck-typed array suitable for use with xarray.
 
-class JaxArrayWrapper(np.lib.mixins.NDArrayOperatorsMixin):
-  """Wraps a JAX array into a duck-typed array suitable for use with xarray.
-
-  This uses an older duck-typed array protocol based on __array_ufunc__ and
-  __array_function__ which works with numpy and xarray. (In newer versions
-  of xarray it implements xarray.namedarray._typing._array_function.)
-
-  This is in the process of being superseded by the Python array API standard
-  (https://data-apis.org/array-api/latest/index.html), but JAX hasn't
-  implemented it yet. Once they have, we should be able to get rid of
-  this wrapper and use JAX arrays directly with xarray.
 
+  Provides a minimal array-like interface required by xarray's constructors and 
+  raises errors for unsupported operations on non-array data types.
+  
   """
-
-  def __init__(self, jax_array):
-    self.jax_array = jax_array
-
+  
+  def __init__(self, leaf: Any, dims: Tuple[Hashable, ...], shape: Tuple[int, ...]):
+    self._leaf = leaf
+    self._dims = dims
+    self._shape = shape
+    
+    # Determine dtype.
+    if isinstance(leaf, bool):
+      self._dtype = jnp.bool_.dtype
+    elif isinstance(leaf, (int, np.integer)):
+      self._dtype = jnp.int32.dtype
+    elif isinstance(leaf, (float, np.floating)):
+      self._dtype = jnp.float32.dtype
+    elif leaf is None:
+      self._dtype = jnp.object_.dtype
+    else:
+      # Attempt to infer dtype, falling back to object.
+      try:
+        self._dtype = jnp.result_type(leaf)
+      except (TypeError, ValueError):
+        self._dtype = jnp.object_.dtype
+  
   def __array_ufunc__(self, ufunc, method, *args, **kwargs):
-    for x in args:
-      if not isinstance(x, (jax.typing.ArrayLike, type(self))):
-        return NotImplemented
     if method != '__call__':
       return NotImplemented
     try:
       # Get the corresponding jax.numpy function to the NumPy ufunc:
       func = getattr(jnp, ufunc.__name__)
```

Comment:
Any reason to support this here, rather than raise an error like you do below in `__array_function__` ?